src/components/utils/useOutsideClick/useOutsideClick.tsx

```diff
@@ -0,0 +1,40 @@
+import React from 'react';
+
+export interface UseOutsideClickProps<T> {
+    ref: React.RefObject<T>;
+    handler?: () => void;
+}
+
+type UseOutsideClickType = <K extends HTMLElement>(props: UseOutsideClickProps<K>) => void;
+
+/**
+ * Hook for observing clicks outside a given target
+ *
+ * @param ref - purpose of observation
+ * @param handler - callback when a click is triggered outside the observation target
+ *
+ * @return - nothing
+ */
+export const useOutsideClick: UseOutsideClickType = ({ref, handler}) => {
+    React.useEffect(() => {
+        if (ref) {
+            const callback = (e: MouseEvent | TouchEvent) => {
+                const elem = ref?.current;
+
+                if (elem && !elem.contains(e.target as Node) && handler) {
+                    handler();
+                }
+            };
+
+            window.addEventListener('click', callback, {capture: true});
+            window.addEventListener('touchstart', callback, {capture: true});
```

Comment:
Shouldn't it be `touchend` event?

Reply:
It depends on how we decide. I assumed that we need a callback to be triggered as soon as the user touches another place. If you set `touchend`, it will work only when the user releases his finger, it may take some time before that

Reply:
But we use click event not mousedown (click happens after mouseup), there should be some consistency

Reply:
I did some research, most popular UI libraries either use click/touchend event or let specify event. For example, [MUI](https://github.com/mui/material-ui/blob/master/packages/mui-base/src/ClickAwayListener/ClickAwayListener.tsx#L237), [ChakraUI](https://github.com/chakra-ui/chakra-ui/blob/main/packages/hooks/use-outside-click/src/index.ts#L66)